io-engine/src/bdev/nexus/nexus_io.rs

```diff
@@ -173,59 +246,58 @@ impl<'n> NexusBio<'n> {
         child: &dyn BlockDevice,
         status: IoCompletionStatus,
     ) {
-        let success = status == IoCompletionStatus::Success;
+        #[cfg(feature = "fault_injection")]
+        let status = self.inject_completion_error(child, status);
 
         debug_assert!(self.ctx().in_flight > 0);
         self.ctx_mut().in_flight -= 1;
 
-        if success {
-            self.ok_checked();
+        if status == IoCompletionStatus::Success {
+            self.ctx_mut().successful += 1;
         } else {
-            // IO failure, mark the IO failed and take the child out
-            error!(
-                "{:?}: IO completion for '{}' failed: {:?}, ctx={:?}",
-                self,
-                child.device_name(),
-                status,
-                self.ctx()
-            );
             self.ctx_mut().status = IoStatus::Failed;
-            self.ctx_mut().must_fail = true;
-            self.handle_failure(child, status);
+            self.ctx_mut().failed += 1;
+
+            self.completion_error(child, status);
         }
-    }
 
-    /// Complete the IO marking at as successfully when all child IO's have been
-    /// accounted for. Failing to account for all child IO's will result in
-    /// a lockup.
-    #[inline]
-    fn ok_checked(&mut self) {
-        if self.ctx().in_flight == 0 {
-            if self.ctx().must_fail {
-                //warn!(?self, "resubmitted due to must_fail");
-                self.retry_checked();
-                //self.fail();
-            } else {
-                self.ok();
-            }
+        if self.ctx().in_flight > 0 {
+            // More child I/Os to complete, not yet ready to complete nexus I/O.
+            trace_io!("Inflight: {self:?}");
+            return;
         }
-    }
 
-    /// Complete the IO marking it as failed.
-    #[inline]
-    fn fail_checked(&mut self) {
-        if self.ctx().in_flight == 0 {
+        if self.ctx().failed == 0 {
+            // No child failures, complete nexus I/O with success.
+            trace_io!("Success: {self:?}");
+            self.ok();
+        } else if self.ctx().successful > 0 {
+            // Having some child failures, resubmit the I/O.
+            self.resubmit();
+        } else {
+            error!("{self:?}: failing nexus I/O: all child I/Os failed");
             self.fail();
         }
     }
 
-    /// retry this IO when all other IOs have completed
-    #[inline]
-    fn retry_checked(&mut self) {
-        if self.ctx().in_flight == 0 {
-            debug!(?self, "resubmitting IO");
-            self.clone().submit_request();
-        }
+    /// Resubmits the I/O.
+    fn resubmit(&mut self) {
+        warn!("{self:?}: resubmitting nexus I/O due to a child I/O failure");
```

Comment:
Should this be debug as we're in the io path? This is a blocking call.

Reply:
We had this warning in resubmit code path before.

Reply:
Interesting.. I think we should modify this (another PR) as it doesn't seem reasonable to have IO waiting on stdout?